### PR TITLE
feat: support SAML POST binding

### PR DESCRIPTION
**Pull request by dacongda**
Fix: #2643 

support saml http post binding when sp use post method

3407f30a1d0fe7eb885472b57577606c

### CODE REVIEW

File: web/src/ApplicationEditPage.js [hunk 1]
```diff
@@ -663,6 +664,17 @@ class ApplicationEditPage extends React.Component {
             }} />
           </Col>
         </Row>
+        <Row style={{marginTop: "20px"}} >
+          <Col style={{marginTop: "5px"}} span={(Setting.isMobile()) ? 19 : 2}>
+            {Setting.getLabel(i18next.t("application:Enable SAML Post Binding"), i18next.t("application:Enable SAML Post Binding - Tooltip"))} :
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:669]: Enable SAML POST binding

File: controllers/saml.go [hunk 2]
```diff
@@ -34,7 +34,12 @@ func (c *ApiController) GetSamlMeta() {
 		return
 	}
 
-	metadata, err := object.GetSamlMeta(application, host)
+	enablePostBinding := c.Input().Get("post")
+	if enablePostBinding == "" {
+		enablePostBinding = "true"
```
Comment (hsluoyz) [inline controllers/saml.go:39]: Why default is true? As we agreed, it's false by default

File: controllers/saml.go [hunk 3]
```diff
@@ -43,3 +48,17 @@ func (c *ApiController) GetSamlMeta() {
...
 }
+
+func (c *ApiController) HandleSamlRedirect() {
+	host := c.Ctx.Request.Host
+
+	owner := c.Ctx.Input.Param(":owner")
+	application := c.Ctx.Input.Param(":application")
+
+	relayState := c.Input().Get("RelayState")
+	samlResponse := c.Input().Get("SAMLRequest")
```
Comment (hsluoyz) [inline controllers/saml.go:59]: Is this `SAMLRequest` spelling correct?

File: controllers/saml.go [hunk 4]
```diff
@@ -34,7 +34,12 @@ func (c *ApiController) GetSamlMeta() {
 		return
 	}
 
-	metadata, err := object.GetSamlMeta(application, host)
+	enablePostBinding := c.Input().Get("post")
+	if enablePostBinding == "" {
```
Comment (hsluoyz) [inline controllers/saml.go:38]: Define `enablePostBinding` as boolean

File: controllers/saml.go [hunk 5]
```diff
@@ -43,3 +48,17 @@ func (c *ApiController) GetSamlMeta() {
...
+
+	owner := c.Ctx.Input.Param(":owner")
+	application := c.Ctx.Input.Param(":application")
+
+	relayState := c.Input().Get("RelayState")
+	samlResponse := c.Input().Get("SAMLRequest")
+
+	targetURL := object.GetSamlRedirectAddress(owner, application, relayState, samlResponse, host)
+
+	c.Redirect(targetURL, 303)
```
Comment (hsluoyz) [inline controllers/saml.go:63]: Don't use int value 303

File: object/saml_idp.go [hunk 6]
```diff
@@ -442,3 +449,8 @@ func NewSamlResponse11(user *User, requestID string, host string) *etree.Element
 
 	return samlResponse
 }
+
+func GetSamlRedirectAddress(owner string, application string, relayState string, samlRequest string, host string) string {
+	originF, _ := getOriginFromHost(host)
```
Comment (hsluoyz) [inline object/saml_idp.go:454]: Handle error

File: web/src/ApplicationEditPage.js [hunk 7]
```diff
@@ -186,8 +187,8 @@ class ApplicationEditPage extends React.Component {
       });
   }
 
-  getSamlMetadata() {
-    ApplicationBackend.getSamlMetadata("admin", this.state.applicationName)
+  getSamlMetadata(checked) {
+    ApplicationBackend.getSamlMetadata("admin", this.state.applicationName, checked ?? false)
```
Comment (hsluoyz) [inline web/src/ApplicationEditPage.js:191]: Just use `checked`

File: web/src/backend/ApplicationBackend.js [hunk 8]
```diff
@@ -89,8 +89,8 @@ export function deleteApplication(application) {
   }).then(res => res.json());
 }
 
-export function getSamlMetadata(owner, name) {
-  return fetch(`${Setting.ServerUrl}/api/saml/metadata?application=${owner}/${encodeURIComponent(name)}`, {
+export function getSamlMetadata(owner, name, enablePostBinding) {
+  return fetch(`${Setting.ServerUrl}/api/saml/metadata?application=${owner}/${encodeURIComponent(name)}&post=${enablePostBinding}`, {
```
Comment (hsluoyz) [inline web/src/backend/ApplicationBackend.js:93]: `&enablePostBinding=${xxx}`

File: web/src/locales/es/data.json [hunk 9]
```diff
@@ -36,6 +36,8 @@
     "Enable SAML C14N10 - Tooltip": "Enable SAML C14N10 - Tooltip",
     "Enable SAML compression": "Activar la compresión SAML",
     "Enable SAML compression - Tooltip": "Si comprimir o no los mensajes de respuesta SAML cuando se utiliza Casdoor como proveedor de identidad SAML",
+    "Enable SAML Post Binding": "Habilitar el enlace SAML POST",
```
Comment (hsluoyz) [inline web/src/locales/es/data.json:39]: No need to translate langs other than zh

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by tiny-v** [comment]
hi, 

i'm using Azure as the SP， Casdoor as the IDP,  and i already enable the Saml Post Binding, but still get the "Unexpected EOF" error when login.

**Comment by tiny-v** [comment]
image

**Comment by hsluoyz** [comment]
@tiny-v create issue and show it's a bug (not your own issue, we don't fix your own issue). Don't comment on old post

**Pull request merged**